massenergize/api
# Notify with sockets

## Related Ticket: https://github.com/massenergize/frontend-admin/issues/947
## Works with frontend admin PR: https://github.com/massenergize/frontend-admin/pull/949

### What happened here? 

- [X] I installed `channels 3.0.5` (its available in requirements.txt) NB: It needs to be version 3.0.5. 
- [X] Switched our app from `wsgi` app to `asgi` app. The changes are in the `settings.py, asgi.py` files i checked in. 
- [X] Created a new app `socket_notifications`. Thats where we can handle all socket routes and consumer classes we need for now and in future. 
- [X] Added one new socket route `ws/me-client/connect/`. This is the route the frontend must hit when the app is loading to setup the connection between the frontend and backend.  
- [X] I also updated `/auth.login` for admins to calculate and record the expiration time of the particular user in session. So that it can be available to our socket consumers to use to track.

### How to test 
Check inside `user_session_tracker_consumer.py` file. Inside the `track` function, there is a commented line that says "FOR TESTING". 
You can uncomment that and it will forcefully expire  your admin in 15 seconds (You can change this during testing if you want). Just so you can have a feel of the flow of everything between the F.E and the B.E when the events are actually triggered.

### Works with frontend admin PR: https://github.com/massenergize/frontend-admin/pull/949
